doc/api/esm.md

```diff
@@ -395,6 +357,322 @@ package in use in an application, which can cause a number of bugs.
+these conditions. For example:
+
+<!-- eslint-skip -->
+```js
+// ./node_modules/pkg/package.json
+{
+  "type": "module",
+  "main": "./pkg.cjs",
+  "exports": {
+    ".": {
+      "module": "./pkg.mjs",
+      "node": "./pkg.cjs"
+    }
+  }
+}
+```
+
+In this example, `require('pkg')` always resolves to `pkg.cjs`, including in
+versions of Node.js where ES modules are unsupported. In Node.js where ES
+modules are supported, `import 'pkg'` references `pkg.mjs`.
+
+The potential for bugs comes from the fact that the `pkg` created by `const pkg
+= require('pkg')` is not the same as the `pkg` created by `import pkg from
+'pkg'`. This is the “divergent specifier hazard,” where one specifer (`'pkg'`)
+resolves to separate files (`pkg.cjs` and `pkg.mjs`) in separate module systems,
+yet both versions might get loaded within an application because Node.js
+supports intermixing CommonJS and ES modules.
+
+Looked at another way, `import pkg from 'pkg'` is a shorthand for `import pkg
+from './node_modules/pkg/pkg.mjs'` and `const pkg = require('pkg')` is a
+shorthand for `const pkg = require('./node_modules/pkg/pkg.cjs')`. Because the
+file paths in the two statements are different, the two `pkg` singletons are
+different.
+
+An `instanceof` comparison of the two returns `false`, and properties added to
+one (like `pkg.foo = 3`) are not present on the other. One ES module file can
+have `import a from 'pkg'` and another can have `import b from 'pkg'`, and `a
+instanceof b` returns `true`. That would not be the case when comparing the `a`
+from `import a from 'pkg'` with the `b` from `const b = require('pkg')`. This
+differs from how `import` and `require` statements work in all-CommonJS or
+all-ES module environments, respectively, and therefore is surprising to users.
+It also differs from the behavior users have grown familiar with over the last
+several years when using transpilation via tools like Babel or
+[`esm`][].
+
+It’s not enough to refer to `'pkg'` using only `require` or only `import` within
+an application; all of that application’s dependencies also need to use the same
+method or the hazard is still present. For example, if an application uses `pkg`
+and `pkg-plugin`, and the application references `pkg` via `import` and
+`pkg-plugin` references `pkg` via `require`, both versions of `pkg` are
+therefore loaded.
+
+#### Writing Dual Packages While Avoiding or Minimizing Hazards
+
+First, the hazard described in the previous section only occurs when a package
+contains both CommonJS and ES module sources and both sources are provided for
+use in Node.js, either via separate main entry points or exported paths. A
+package could instead be written where any version of Node.js receives only
+CommonJS sources, and any separate ES module sources the package may contain
+could be intended only for other environments such as browsers. Such a package
+would be usable by any version of Node.js, since `import` can refer to CommonJS
+files.
```

Comment:
Perhaps rather just split into three approaches:

1. Ship CommonJS (and the benefits of this)
2. ESM Wrappers for CommonJS
3. Concepts of state isolation

The order of going from most applicable and general to more specific handling works well. And that is the same as what you have, but the amount of material still needs to be cut down considerably.